src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReservoirDownsampler.java

```diff
@@ -110,35 +120,39 @@ public void submit ( final GATKRead newRead ) {
 
     @Override
     public boolean hasFinalizedItems() {
-        return ! reservoir.isEmpty();
+        // All items in the reservoir are pending until endOfInputStream is seen, at which point all items become finalized
+        return endOfInputStream && !reservoir.isEmpty();
     }
 
     @Override
     public List<GATKRead> consumeFinalizedItems() {
+        Utils.validate(endOfInputStream == true, "signalEndOfInput must be called before finalized items can be consumed");
```

Comment:
If `consumeFinalizedItems()` is called when there are no finalized items, it should return an empty list, not throw an exception.

Reply:
Are you suggesting that this `validate` call be removed ? It just validates that `signalEndOfInput` has been called, which is a precondition for having finalized items. To me, hitting this exception indicates a contract violation - the caller shouldn't expect there to be finalized items until eoi is called. This caught the case in MutectDownsampler where it was calling` submit` followed by `consume` with no intervening `signalEndOfInput`.

Reply:
Yes, I think it should be removed. The contract of `consumeFinalizedItems()` is that it should return an empty list if there are no finalized items (without any side effects such as resetting the downsampler) -- this check violates that contract by throwing an exception instead.